### ISSUE TITLE
Comparing Records with polymorphic member fails

### ISSUE THREAD

**Issue by slimshader** [louthy/language-ext#559]
When comparing Record<> with polymorphic member (which simulates Discriminated Union), comparison fails with cast exception. Is that the desired behavior? How to handle comparison in such cases? I hoped it would return false in case of different "union case".

Also, I've noticed different behavior when such Records are stored in Map vs direct compare or Lst compare. Maps do not throw, while direct and Lst comparison thows cast exception.

Code to reproduce:
```
using LanguageExt;

namespace ConsoleApp4
{
  interface ArmyState
  {

  }

  class Idle : Record<Idle>, ArmyState
  {
    public Idle(int loc)
    {
      Location = loc;
    }

    public int Location;
  }

  class Dead : Record<Idle>, ArmyState
  {
    public int Location;

    public Dead(int loc)
    {
      Location = loc;
    }
  }

  class Army : Record<Army>
  {
    public int Id { get; }
    public ArmyState State { get; }

    public Army(int id)
    {
      Id = id;
      State = new Idle(5);
    }

    public int Location
    {
      get
      {
        switch (State)
        {
          case Idle idle: return idle.Location;
          case Dead dead: return dead.Location;
          default: return -1;
        }
      }
    }

    private Army(int id, ArmyState state)
    {
      Id = id;
      State = state;
    }

    public Army ToIdle() => new Army(Id, new Idle(Location));
    public Army ToDead() => new Army(Id, new Dead(Location));
  }

  class Program
  {
    static void Main(string[] args)
    {
      var a1 = new Army(1).ToDead();
      var a2 = new Army(1);

      var m1 = new Map<int, Army>().Add(a1.Id, a1);
      var m2 = new Map<int, Army>().Add(a2.Id, a2);


      var mapEq = m1 == m2; // No Exception, true

      var eq = a1 == a2; // No excpetion

      var lstEq = m1.Freeze() == m2.Freeze(); // Exception
    }
  }
}
```

**Comment by colethecoder** [comment]
I've just tried your sample with the latest nuget package and I found this line throws an InvalidCastException:

```cs
var eq = a1 == a2; 
```

but I believe this is because you have:

```cs
class Dead : Record<Idle>, ArmyState
```

which I think should be:

```cs
class Dead : Record<Dead>, ArmyState
```

Adjusting that I then get:

```cs
var mapEq = m1 == m2;               // No Exception, true
var eq = a1 == a2;                  // No Exception, false
var lstEq = 
    m1.Freeze() == m2.Freeze();     // No Exception, false
```

So map equality is a bit weird but the rest seems to be as you would expect.

**Comment by bender2k14** [comment]
> ...you have:
> 
> ```cs
> class Dead : Record<Idle>, ArmyState
> ```
> 
> which I think should be:
> 
> ```cs
> class Dead : Record<Dead>, ArmyState
> ```

That is such an interesting typo.  I was thinking that it should be a compile error because `Record<T>` has the type constrained `T : Record<T>`.  The intention is that `T` is the type that is extending `Record<T>`.  But as this example shows, the constraint is weaker than the intention because
```
class Idle : Record<Idle>, ArmyState { ... }
class Dead : Record<Idle>, ArmyState { ... }
```

together satisfy both constraints.

**Comment by bender2k14** [comment]
> Adjusting that I then get:
> 
> ```cs
> var mapEq = m1 == m2;               // No Exception, true
> var eq = a1 == a2;                  // No Exception, false
> var lstEq = 
>     m1.Freeze() == m2.Freeze();     // No Exception, false
> ```
> 
> So map equality is a bit weird but the rest seems to be as you would expect.
> 
> Edit: Map is only comparing keys which I assume is because the value type doesn't have to be equatable

Yea, I suppose it could go either way.  I created PR #561 that adds equality checking for the values.  I am interested to see what @louthy wants as the intended behavior.

**Comment by slimshader** [comment]
When preparing this test code I was actually trying to reproduce minimal version of the problem in larger code-base. While I've made a bug with Record<T> here, I realized (after your replies) that in my original code my variants didn't derive from Record<> at all.

Still, the main reason I posted the issue was Map vs Lst comparison. I wrongly assumed it was related to polymorphic member in Record. My case is that I had a Lst of Records that was updated in similar fashion to my example (Army is now Dead), in the view layer new model was compared against old model to potentially refresh the view (a'la model-view-update). When using Lst change was detected but when I switched to Map, it no longer shows difference between old and updated model.
